File: src/solutions/scala/N867TransposeMatrix.scala
package solutions.scala

object N867TransposeMatrix {

  def transpose(A: Array[Array[Int]]): Array[Array[Int]] = A(0).indices.map(col => A.map(row => row(col))).toArray

  def transpose2(A: Array[Array[Int]]): Array[Array[Int]] = A.transpose

  def main(args: Array[String]): Unit = {
    assert(transpose(Array(Array(1, 2, 3), Array(4, 5, 6), Array(7, 8, 9)))
      .sameElements(Array(Array(1, 4, 7), Array(2, 5, 8), Array(3, 6, 9)))
    )
    assert(transpose(Array(Array(1, 2, 3), Array(4, 5, 6)))
      .sameElements(Array(Array(1, 4), Array(2, 5), Array(3, 6)))
    )
  }
}

File: src/solutions/java/N867TransposeMatrix.java
package solutions.java;

public class N867TransposeMatrix {

    public int[][] transpose(int[][] a) {
        int rowNum = a.length;
        int colNum = a[0].length;
        int[][] res = new int[colNum][rowNum];
        for (int r = 0; r < rowNum; r++) {
            for (int c = 0; c < colNum; c++) {
                res[c][r] = a[r][c];
            }
        }
        return res;
    }
}
